NMS-16259: Metadata manipulation via rest (#6937)
OpenNMS/opennms: docs/modules/development/pages/rest/meta-data.adoc

[[metadata-rest]]
= Metadata

You can query, add, modify, or delete the actual metadata of nodes, interfaces, and services via REST endpoints.

Modification of metadata entries is limited to user defined contexts (starting with `X-`).

> Read about xref:operation:deep-dive/admin/metadata.adoc#metadata-contexts[how Metadata context work] for further details.

Reviewer: ```suggestion
Read about xref:operation:deep-dive/admin/metadata.adoc#metadata-contexts[how metadata contexts work] for further details.
```

WARNING: It is highly recommended to not modify any metadata using this interface but use the provided requisition mechanisms instead.
Altering metadata from a node configured in a requisition will change your existing metadata on every synchronisation of the originating requisition in an unpredicted way.

> This can lead to data-lose. You have been warned!

Reviewer 2: ```suggestion
This can lead to data-loss. You have been warned!
```

> Altering metadata from a node configured in a requisition will change your existing metadata on every synchronisation of the originating requisition in an unpredicted way.
> This can lead to data-loss. You have been warned!

Reviewer: ```suggestion
Nodes provisioned via requisition may have metadata overwritten during the resynchronization process.
This can lead to loss of custom metadata defined outside the requisition.
```

| api/v2/nodes/\{id}/ipinterfaces/\{ipinterface}/services/\{service}/metadata/\{context}/\{key}
| Get the entry for the given context and key associated with this service.
|===

[[rest-api-meta-data-post]]
== POSTs (Adding Meta-Data)

POST requires XML using `application/xml` as its Content-Type.

> [options="header", cols="5,10"]
> |===
> | Resource                                                            | Description
> | `api/v2/nodes/\{id\}/metadata`                                      | Adds a metadata entry to the given node.
> | `api/v2/nodes/\{id\}/\{ipinterface\}/metadata`                      | Adds a metadata entry to the given interface.
> | `api/v2/nodes/\{id\}/\{ipinterface\}/services/\{service\}/metadata` | Adds a metadata entry to the given service.
> |===

Reviewer: ```suggestion
[options="header", cols="5,10"]
|===
| Resource  | Description

| api/v2/nodes/\{id\}/metadata
| Adds a metadata entry to the given node.
| api/v2/nodes/\{id\}/\{ipinterface\}/metadata
| Adds a metadata entry to the given interface.

| api/v2/nodes/\{id\}/\{ipinterface\}/services/\{service\}/metadata`
| Adds a metadata entry to the given service.
|===
```
